README.md
# Discode

A Visual Studio Code extension that allows you to share code to Discord with webhooks.

*[Make sure to join the official Discode Discord server here!](https://discord.gg/Cm3ADqfPpp)*

## Usage

After installing the extension, open up the Command Palette and under "Discode" use the "Set a Webhook" command.
This command can also be used to update an existing webhook, and while you can configure it in the settings it is generally recommended to do so with the command.

![settingUp](images/settingUp.png)

After you're done setting up, all it takes is to select text, right click and press "Share Code to Discord"!

![usage](images/usage.png)

![result](images/result.png)

*You can also use the same command through the Command Palette.*

## Extension Settings

This extension contributes the following settings:

* `discode.webhook`: The webhook the extension uses to share the code to.
* `discode.webhookSecond`: An additional second webhook.
* `discode.webhookThird`: An additional third webhook.
* `discode.name`: The name that gets set as the embed's author.
* `discode.avatar`: The image that gets set as the embed's author's avatar.

## Known Issues

* None so far!

## Release Notes

### Release 2.0.1

* Added additional two webhook slots.
* Removed the "Community Projects" section from the README.

### Release 2.0.0

* Major code and file restructuring.
* Re-added support for generating Hastebin links for longer code.

### Release 1.1.3

* Added support for Discord Canary and PTB webhook links.
* Added "Community Projects" section in the README.md file.

### Release 1.1.2

* Removed Hastebin share feature due to outdated and incompatible modules.
* Upped long code length limit to Discord's maximum message length limit (2000 characters).
* Embed now shows the file name above the code.

### Release 1.1.1

* Hotfix for Release 1.1.0 throwing an error when attempting to share code.

### Release 1.1.0

* "Set a Webhook" command's input boxes don't disappear now if window focus is lost.
* Code longer than 1024 characters is now sent through a Hastebin link.
* Unnecessary whitespace is now removed from selection.
* The embed's footer now has Discode's logo.
* Grammatical fixes.

### Release 1.0.0

* Initial release of Discode.


CHANGELOG.md
## Release 2.0.1

- Added additional two webhook slots.
- Removed the "Community Projects" section from the README.

## Release 2.0.0

- Major code and file restructuring.
- Re-added support for generating Hastebin links for longer code.

## Release 1.1.3

- Added support for Discord Canary and PTB webhook links.
- Added "Community Projects" section in the README.md file.

## Release 1.1.2

- Removed Hastebin share feature due to outdated and incompatible modules.
- Upped long code length limit to Discord's maximum message length limit (2000 characters).
- Embed now shows the file name above the code.

## Release 1.1.1

- Hotfix for Release 1.1.0 throwing an error when attempting to share code.

## Release 1.1.0

- "Set a Webhook" command's input boxes don't disappear now if window focus is lost.
- Code longer than 1024 characters is now sent through a Hastebin link.
- Unnecessary whitespace is now removed from selection.
- The embed's footer now has Discode's logo.
- Grammatical fixes.

## Release 1.0.0

- Initial release.


src/extension.ts
import * as vscode from 'vscode';
import * as eol from 'eol';
import { Webhook, MessageBuilder } from 'discord-webhook-node';
const hastebin = require('hastebin-save');

export async function activate(context: vscode.ExtensionContext) {
	console.log('The "Discode" extension has been loaded successfully.');

	async function setWebhook(count = 1) {
		const webhookInput = await vscode.window.showInputBox({
			ignoreFocusOut: true,
			placeHolder: 'Input a valid Discord webhook link.',
			validateInput: text => {
				var expr = /(https?):\/\/((?:ptb\.|canary\.)?discord(?:app)?\.com)\/api(?:\/)?(v\d{1,2})?\/webhooks\/(\d{17,19})\/([\w\-]{68})/i;
				return expr.test(text) ? null : 'That is not a valid Discord webhook link.';
			}
		});

		if (count === 1) {
			const nameInput = await vscode.window.showInputBox({
				ignoreFocusOut: true,
				placeHolder: 'Input a name to show in the embed. Make sure it\'s less than 32 characters.',
				validateInput: text => {
					return text.length <= 32 ? null : 'That name is too long.';
				}
			});
	
			const avatarInput = await vscode.window.showInputBox({
				ignoreFocusOut: true,
				placeHolder: 'Input an image link to show in the embed as the avatar. Supported formats: png, jpg, gif.',
				validateInput: text => {
					return text.endsWith('png') || text.endsWith('jpeg') || text.endsWith('jpg') || text.endsWith('gif') ? null : 'That format is not supported.';
				}
			});

			if (webhookInput) {
				await vscode.workspace.getConfiguration('discode').update('webhook', webhookInput, vscode.ConfigurationTarget.Global);
					
				if (nameInput) {
					await vscode.workspace.getConfiguration('discode').update('name', nameInput, vscode.ConfigurationTarget.Global);
				}
	
				if (avatarInput) {
					await vscode.workspace.getConfiguration('discode').update('avatar', avatarInput, vscode.ConfigurationTarget.Global);
				}
			}
		} else {
			if (webhookInput) {
				switch (count) {
					case 2:
						await vscode.workspace.getConfiguration('discode').update('webhookSecond', webhookInput, vscode.ConfigurationTarget.Global);
						break;
					case 3:
						await vscode.workspace.getConfiguration('discode').update('webhookThird', webhookInput, vscode.ConfigurationTarget.Global);
						break;
				}
			}
		}

		await vscode.window.showInformationMessage('Successfully set a webhook!');
	}

	async function shareCode(count = 1) {
		const errorMessage = 'Discode has not been set up! You may be missing a name or avatar. Please use the "Set a Webhook" command through the Command Palette or configure it manually in the settings.';

		switch (count) {
			case 1:
				if (!vscode.workspace.getConfiguration('discode').get('webhook') || !vscode.workspace.getConfiguration('discode').get('name')) {
					return vscode.window.showErrorMessage(errorMessage);
				}
				break;
			case 2:
				if (!vscode.workspace.getConfiguration('discode').get('webhookSecond') || !vscode.workspace.getConfiguration('discode').get('name')) {
					return vscode.window.showErrorMessage(errorMessage);
				}
				break;

			case 3:
				if (!vscode.workspace.getConfiguration('discode').get('webhookThird') || !vscode.workspace.getConfiguration('discode').get('name')) {
					return vscode.window.showErrorMessage(errorMessage);
				}
				break;
		}

		const editor = vscode.window.activeTextEditor;
		
		if (editor) {
			const document = editor.document;
			const selection = editor.selection;
			const text = document.getText(selection);

			let currentHook: string;

			switch (count) {
				case 1:
					currentHook = `${vscode.workspace.getConfiguration('discode').get('webhook')}`;
					break;
				case 2:
					currentHook = `${vscode.workspace.getConfiguration('discode').get('webhookSecond')}`;
					break;
				case 3:
					currentHook = `${vscode.workspace.getConfiguration('discode').get('webhookThird')}`;
					break;
			}

			const hook = new Webhook({
				url: currentHook!,
				retryOnLimit: false
			});

			const whitespaceAmount = text.length - text.trimLeft().length;
			const codeLines = eol.split(text);

			let code: string[] = [];

			codeLines.forEach(function(line: string) {
				code.push(line.substring(whitespaceAmount));
			});
			
			if (`${code}`.length > 2048) {
				try {
					let hastebinLink : string;

					// @ts-ignore
					await hastebin.upload(`${code.join('\n')}`, link => {
						hastebinLink = `https://hastebin.com/${link}`;
					});

					vscode.window.showInformationMessage('Please wait...');

					setTimeout(async function() {
						let embed = new MessageBuilder()
							.setColor(0x27b5f4)
							.setDescription(`**${document.fileName.substring(document.fileName.lastIndexOf('\\') + 1)}**\n${hastebinLink}`)
							.setFooter('Sent with Discode from Visual Studio Code', 'https://cdn.discordapp.com/attachments/798580610686779392/798580822641868870/embedIcon.png')
							.setTimestamp();
		
						if (!vscode.workspace.getConfiguration('discode').get('avatar')) {
							embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`);
						} else {
							embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`, `${vscode.workspace.getConfiguration('discode').get('avatar')}`);
						}

						await hook.send(embed);

						await vscode.window.showInformationMessage('Successfully shared code to Discord!');
					}, 3000);
				} catch (error) {
					console.log(error);
					await vscode.window.showErrorMessage('The webhook configuration is invalid! Please check if everything is correct in the settings or use the "Set a Webhook" command through the Command Palette.');
				}
			} else {
				try {
					let embed = new MessageBuilder()
						.setColor(0x27b5f4)
						.setDescription(`**${document.fileName.substring(document.fileName.lastIndexOf('\\') + 1)}**\`\`\`${document.languageId}\n${code.join('\n')}\`\`\``)
						.setFooter('Sent with Discode from Visual Studio Code', 'https://cdn.discordapp.com/attachments/798580610686779392/798580822641868870/embedIcon.png')
						.setTimestamp();
		
					if (!vscode.workspace.getConfiguration('discode').get('avatar')) {
						embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`);
					} else {
						embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`, `${vscode.workspace.getConfiguration('discode').get('avatar')}`);
					}

					await hook.send(embed);

					await vscode.window.showInformationMessage('Successfully shared code to Discord!');
				} catch (error) {
					console.log(error);
					await vscode.window.showErrorMessage('The webhook configuration is invalid! Please check if everything is correct in the settings or use the "Set a Webhook" command through the Command Palette.');
				}
			}
		}
	}

	let set = vscode.commands.registerCommand('discode.set', async() => {
		await setWebhook(1);
	});

	let setSecond = vscode.commands.registerCommand('discode.setSecond', async() => {
		await setWebhook(2);
	});
	let setThird = vscode.commands.registerCommand('discode.setThird', async() => {
		await setWebhook(3);
	});

	let share = vscode.commands.registerCommand('discode.share', async() => {
		await shareCode(1);
	});

	let shareSecond = vscode.commands.registerCommand('discode.shareSecond', async() => {
		await shareCode(2);
	});

	let shareThird = vscode.commands.registerCommand('discode.shareThird', async() => {
		await shareCode(3);
	});

	context.subscriptions.push(set);
	context.subscriptions.push(setSecond);
	context.subscriptions.push(setThird);
	context.subscriptions.push(share);
	context.subscriptions.push(shareSecond);
	context.subscriptions.push(shareThird);
}

export function deactivate() {}


out/extension.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const eol = require("eol");
const discord_webhook_node_1 = require("discord-webhook-node");
const hastebin = require('hastebin-save');
function activate(context) {
    return __awaiter(this, void 0, void 0, function* () {
        console.log('The "Discode" extension has been loaded successfully.');
        function setWebhook(count = 1) {
            return __awaiter(this, void 0, void 0, function* () {
                const webhookInput = yield vscode.window.showInputBox({
                    ignoreFocusOut: true,
                    placeHolder: 'Input a valid Discord webhook link.',
                    validateInput: text => {
                        var expr = /(https?):\/\/((?:ptb\.|canary\.)?discord(?:app)?\.com)\/api(?:\/)?(v\d{1,2})?\/webhooks\/(\d{17,19})\/([\w\-]{68})/i;
                        return expr.test(text) ? null : 'That is not a valid Discord webhook link.';
                    }
                });
                if (count === 1) {
                    const nameInput = yield vscode.window.showInputBox({
                        ignoreFocusOut: true,
                        placeHolder: 'Input a name to show in the embed. Make sure it\'s less than 32 characters.',
                        validateInput: text => {
                            return text.length <= 32 ? null : 'That name is too long.';
                        }
                    });
                    const avatarInput = yield vscode.window.showInputBox({
                        ignoreFocusOut: true,
                        placeHolder: 'Input an image link to show in the embed as the avatar. Supported formats: png, jpg, gif.',
                        validateInput: text => {
                            return text.endsWith('png') || text.endsWith('jpeg') || text.endsWith('jpg') || text.endsWith('gif') ? null : 'That format is not supported.';
                        }
                    });
                    if (webhookInput) {
                        yield vscode.workspace.getConfiguration('discode').update('webhook', webhookInput, vscode.ConfigurationTarget.Global);
                        if (nameInput) {
                            yield vscode.workspace.getConfiguration('discode').update('name', nameInput, vscode.ConfigurationTarget.Global);
                        }
                        if (avatarInput) {
                            yield vscode.workspace.getConfiguration('discode').update('avatar', avatarInput, vscode.ConfigurationTarget.Global);
                        }
                    }
                }
                else {
                    if (webhookInput) {
                        switch (count) {
                            case 2:
                                yield vscode.workspace.getConfiguration('discode').update('webhookSecond', webhookInput, vscode.ConfigurationTarget.Global);
                                break;
                            case 3:
                                yield vscode.workspace.getConfiguration('discode').update('webhookThird', webhookInput, vscode.ConfigurationTarget.Global);
                                break;
                        }
                    }
                }
                yield vscode.window.showInformationMessage('Successfully set a webhook!');
            });
        }
        function shareCode(count = 1) {
            return __awaiter(this, void 0, void 0, function* () {
                const errorMessage = 'Discode has not been set up! You may be missing a name or avatar. Please use the "Set a Webhook" command through the Command Palette or configure it manually in the settings.';
                switch (count) {
                    case 1:
                        if (!vscode.workspace.getConfiguration('discode').get('webhook') || !vscode.workspace.getConfiguration('discode').get('name')) {
                            return vscode.window.showErrorMessage(errorMessage);
                        }
                        break;
                    case 2:
                        if (!vscode.workspace.getConfiguration('discode').get('webhookSecond') || !vscode.workspace.getConfiguration('discode').get('name')) {
                            return vscode.window.showErrorMessage(errorMessage);
                        }
                        break;
                    case 3:
                        if (!vscode.workspace.getConfiguration('discode').get('webhookThird') || !vscode.workspace.getConfiguration('discode').get('name')) {
                            return vscode.window.showErrorMessage(errorMessage);
                        }
                        break;
                }
                const editor = vscode.window.activeTextEditor;
                if (editor) {
                    const document = editor.document;
                    const selection = editor.selection;
                    const text = document.getText(selection);
                    let currentHook;
                    switch (count) {
                        case 1:
                            currentHook = `${vscode.workspace.getConfiguration('discode').get('webhook')}`;
                            break;
                        case 2:
                            currentHook = `${vscode.workspace.getConfiguration('discode').get('webhookSecond')}`;
                            break;
                        case 3:
                            currentHook = `${vscode.workspace.getConfiguration('discode').get('webhookThird')}`;
                            break;
                    }
                    const hook = new discord_webhook_node_1.Webhook({
                        url: currentHook,
                        retryOnLimit: false
                    });
                    const whitespaceAmount = text.length - text.trimLeft().length;
                    const codeLines = eol.split(text);
                    let code = [];
                    codeLines.forEach(function (line) {
                        code.push(line.substring(whitespaceAmount));
                    });
                    if (`${code}`.length > 2048) {
                        try {
                            let hastebinLink;
                            // @ts-ignore
                            yield hastebin.upload(`${code.join('\n')}`, link => {
                                hastebinLink = `https://hastebin.com/${link}`;
                            });
                            vscode.window.showInformationMessage('Please wait...');
                            setTimeout(function () {
                                return __awaiter(this, void 0, void 0, function* () {
                                    let embed = new discord_webhook_node_1.MessageBuilder()
                                        .setColor(0x27b5f4)
                                        .setDescription(`**${document.fileName.substring(document.fileName.lastIndexOf('\\') + 1)}**\n${hastebinLink}`)
                                        .setFooter('Sent with Discode from Visual Studio Code', 'https://cdn.discordapp.com/attachments/798580610686779392/798580822641868870/embedIcon.png')
                                        .setTimestamp();
                                    if (!vscode.workspace.getConfiguration('discode').get('avatar')) {
                                        embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`);
                                    }
                                    else {
                                        embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`, `${vscode.workspace.getConfiguration('discode').get('avatar')}`);
                                    }
                                    yield hook.send(embed);
                                    yield vscode.window.showInformationMessage('Successfully shared code to Discord!');
                                });
                            }, 3000);
                        }
                        catch (error) {
                            console.log(error);
                            yield vscode.window.showErrorMessage('The webhook configuration is invalid! Please check if everything is correct in the settings or use the "Set a Webhook" command through the Command Palette.');
                        }
                    }
                    else {
                        try {
                            let embed = new discord_webhook_node_1.MessageBuilder()
                                .setColor(0x27b5f4)
                                .setDescription(`**${document.fileName.substring(document.fileName.lastIndexOf('\\') + 1)}**\`\`\`${document.languageId}\n${code.join('\n')}\`\`\``)
                                .setFooter('Sent with Discode from Visual Studio Code', 'https://cdn.discordapp.com/attachments/798580610686779392/798580822641868870/embedIcon.png')
                                .setTimestamp();
                            if (!vscode.workspace.getConfiguration('discode').get('avatar')) {
                                embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`);
                            }
                            else {
                                embed.setAuthor(`${vscode.workspace.getConfiguration('discode').get('name')}`, `${vscode.workspace.getConfiguration('discode').get('avatar')}`);
                            }
                            yield hook.send(embed);
                            yield vscode.window.showInformationMessage('Successfully shared code to Discord!');
                        }
                        catch (error) {
                            console.log(error);
                            yield vscode.window.showErrorMessage('The webhook configuration is invalid! Please check if everything is correct in the settings or use the "Set a Webhook" command through the Command Palette.');
                        }
                    }
                }
            });
        }
        let set = vscode.commands.registerCommand('discode.set', () => __awaiter(this, void 0, void 0, function* () {
            yield setWebhook(1);
        }));
        let setSecond = vscode.commands.registerCommand('discode.setSecond', () => __awaiter(this, void 0, void 0, function* () {
            yield setWebhook(2);
        }));
        let setThird = vscode.commands.registerCommand('discode.setThird', () => __awaiter(this, void 0, void 0, function* () {
            yield setWebhook(3);
        }));
        let share = vscode.commands.registerCommand('discode.share', () => __awaiter(this, void 0, void 0, function* () {
            yield shareCode(1);
        }));
        let shareSecond = vscode.commands.registerCommand('discode.shareSecond', () => __awaiter(this, void 0, void 0, function* () {
            yield shareCode(2);
        }));
        let shareThird = vscode.commands.registerCommand('discode.shareThird', () => __awaiter(this, void 0, void 0, function* () {
            yield shareCode(3);
        }));
        context.subscriptions.push(set);
        context.subscriptions.push(setSecond);
        context.subscriptions.push(setThird);
        context.subscriptions.push(share);
        context.subscriptions.push(shareSecond);
        context.subscriptions.push(shareThird);
    });
}
exports.activate = activate;
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map